dbsnap_verify/state_doc.py

```diff
@@ -0,0 +1,311 @@
+            self.setattrs_from_dict(document)
+        elif isinstance(document, basestring):
+            self.setattrs_from_dict(json.loads(document))
+        else:
+            raise Exception(
+                "Invalid type ({}), must be dict or JSON basestring.".format(
+                    type(document)
+                )
+            )
+
+    @property
+    def to_json(self):
+        return json.dumps(self.__dict__, indent=2)
+
+
+class StateDoc(DocToObject):
+
+    def __init__(
+            self,
+            name,
+            states=None,
+            state_doc_path=None,
+            state_doc_bucket=None,
+            **kwargs):
+
+        if states is None:
+            self.states = []
+        else:
+            self.states = states
+        self.state_doc_name = name
+        self.state_doc_path = state_doc_path
+        self.state_doc_bucket = state_doc_bucket
+        super(StateDoc, self).__init__(kwargs)
+
+    @property
+    def state_doc_bucket_name(self):
+        return environ.get("STATE_DOC_BUCKET", self.state_doc_bucket)
+
+    @property
+    def state_doc_file_path(self):
+        return environ.get("STATE_DOC_PATH", self.state_doc_path)
+
+    @property
+    def current_state(self):
+        if self.states:
+            return self.states[-1]["state"]
+
+    @property
+    def persistence(self):
+        """str: "state_doc_bucket" or "state_doc_path" or raises exception."""
+        if self.state_doc_bucket_name and self.state_doc_file_path:
+            msg = "Choose either `state_doc_bucket` or `state_doc_path` not both."
+            raise Exception(msg)
+        elif self.state_doc_bucket_name:
+            return "state_doc_bucket"
+        elif self.state_doc_file_path:
+            return "state_doc_path"
+
+    @property
+    def state_doc_s3_key(self):
+        return "state-doc-{}.json".format(self.state_doc_name)
+
+    def _save_state_doc_in_s3(self):
+        if self.state_doc_bucket_name:
+            s3 = boto3.client("s3")
+            s3.put_object(
+                Bucket=self.state_doc_bucket_name,
+                Key=self.state_doc_s3_key,
+                Body=self.to_json,
+            )
+    
+    def _load_state_doc_from_s3(self):
+        """Returns a JSON String State Document."""
+        s3 = boto3.client("s3")
+        s3_object = s3.get_object(
+            Bucket=self.state_doc_bucket_name,
+            Key=self.state_doc_s3_key,
+        )
+        return s3_object["Body"].read().decode("utf-8")
+
+    def _save_state_doc_in_path(self):
+        with open(self.state_doc_file_path, 'w') as json_file:
+            json_file.write(self.to_json)
+
+    def _load_state_doc_from_path(self):
+        """Returns a JSON String State Document."""
+        with open(self.state_doc_file_path, 'r') as json_file:
+            return json_file.read()
+
+    def transition_state(self, new_state):
+        """Change the state of the StateDoc and save to persistence."""
+        self.states.append(
```

Comment:
There's a state machine, but I don't see anywhere where allowed state transitions are defined and validated. (e.g. what are valid states to transition to from`wait`?).

You have a pretty good diagram of the state transitions, but I don't see where that's verified in code (I could be missing it).

Reply:
Please see https://github.com/remind101/dbsnap-verify/pull/2/commits/32b626b7622c2add0e266ab0bba7a429c1f7ac7e